csharp/extractor/Semmle.Extraction.CSharp.Standalone/BuildAnalysis.cs

```diff
@@ -100,28 +71,41 @@ public BuildAnalysis(Options options, IProgressMonitor progress)
                 dllDirNames.Add(runtimeLocation);
             }
 
-            // TODO: remove the below when the required SDK is installed
-            using (new FileRenamer(sourceDir.GetFiles("global.json", SearchOption.AllDirectories)))
+            if (options.UseMscorlib)
             {
-                var solutions = options.SolutionFile is not null ?
-                        new[] { options.SolutionFile } :
-                        sourceDir.GetFiles("*.sln", SearchOption.AllDirectories).Select(d => d.FullName);
+                UseReference(typeof(object).Assembly.Location);
+            }
+
+            packageDirectory = new TemporaryDirectory(ComputeTempDirectory(sourceDir.FullName));
 
-                if (options.UseNuGet)
+            if (options.UseNuGet)
+            {
+                dllDirNames.Add(packageDirectory.DirInfo.FullName);
+                try
                 {
-                    RestoreSolutions(solutions);
+                    var nuget = new NugetPackages(sourceDir.FullName, packageDirectory, progressMonitor);
+                    nuget.InstallPackages();
+                }
+                catch (FileNotFoundException)
+                {
+                    progressMonitor.MissingNuGet();
                 }
-                dllDirNames.Add(packageDirectory.DirInfo.FullName);
-                assemblyCache = new BuildAnalyser.AssemblyCache(dllDirNames, progress);
-                AnalyseSolutions(solutions);
 
-                foreach (var filename in assemblyCache.AllAssemblies.Select(a => a.Filename))
-                    UseReference(filename);
+                // TODO: remove the below when the required SDK is installed
+                using (new FileRenamer(sourceDir.GetFiles("global.json", SearchOption.AllDirectories)))
+                {
+                    Restore(solutions);
+                    Restore(allProjects);
+                    DownloadMissingPackages(allProjects);
+                }
             }
 
-            if (options.UseMscorlib)
+            assemblyCache = new AssemblyCache(dllDirNames, progressMonitor);
+            AnalyseSolutions(solutions);
+
+            foreach (var filename in assemblyCache.AllAssemblies.Select(a => a.Filename))
```

Comment:
Maybe use .ForEach

Reply:
I generally try to avoid `.ForEach` because
- I don't think it helps readability over `foreach`
- It's defined on `List<>` and not `IEnumerable<>`
- I think it performs worse than `foreach`. (I tried to put together a small benchmark [here](https://dotnetfiddle.net/KikmIx). `.ForEach` is a bit slower, and allocates more)

Reply:
Fair enough 👍 
I was also thinking of our own IEnumerable extension method `ForEach`.
Yes, it will be marginally slower for very large collections, but that should be dominated fairly quickly by any computation within the loop.

Reply:
Ahh, I see. We seem to use our own `ForEach` with single line lambdas, where readability might actually be somewhat better then a multi-line `foreach`.